src/main/ruby/core/math.rb

```diff
@@ -15,6 +15,15 @@ module Math
 
   module_function
 
+  def hypot(a, b)
+    if Integer === a and Integer === b
+      # Much faster (~10x) than calling the Math.hypot() / hypot(3)
+      Math.sqrt(a*a + b*b)
+    else
+      Truffle.invoke_primitive :math_hypot, a, b
+    end
+  end
```

Comment:
Why aren't you using a specialisation in Java code here?

Reply:
I need `*` to handle overflow, and the `math_hypot` primitives uses SimpleDyadicMathNode so it's not easy to add extra specializations for `(int, int)` and `(long, long)`.

Reply:
Sometimes I think we should have a special kind of `case` statement that works like our specialisations with the expressions as guards. It would be different from a normal `case` because it would keep using the cases that matched before without trying each guard in order each time.

Reply:
Right, so these `Integer ===` checks would only be on the fast-path if there ever was an Integer.
I think in simple cases like this it might optimize just as well.
For example, if 2 doubles are passed, the integer checks should be removed by Graal as their implementation check the class of the objects and become redundant with the existing check that the arguments are `double`.